File: kotlin/ktor/src/main/kotlin/com/example/petstore/repository/InMemoryPetstoreRepository.kt
package com.example.petstore.repository

import com.example.petstore.models.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

class InMemoryPetstoreRepository : PetstoreRepository {

    private val pets = ConcurrentHashMap<Long, Pet>()
    private val orders = ConcurrentHashMap<Long, Order>()
    private val users = ConcurrentHashMap<String, User>()
    private val photos = ConcurrentHashMap<Long, ByteArray>()

    private val petIdSeq = AtomicLong(0)
    private val orderIdSeq = AtomicLong(0)
    private val photoIdSeq = AtomicLong(0)

    private fun nextPetId() = (pets.keys.maxOrNull() ?: 0L) + 1L
    private fun nextOrderId() = (orders.keys.maxOrNull() ?: 0L) + 1L
    private fun nextPhotoId() = (photos.keys.maxOrNull() ?: 0L) + 1L

    // -------------------------------------------------------------------------
    // Pet
    // -------------------------------------------------------------------------

    override fun addPet(pet: Pet): Pet {
        val id = pet.id ?: nextPetId()
        val saved = pet.copy(id = id)
        pets[id] = saved
        return saved
    }

    override fun updatePet(pet: Pet): Pet {
        val id = pet.id ?: throw InvalidInputException("Pet id required for update")
        pets[id] ?: throw NotFoundException("Pet not found: $id")
        val updated = pet.copy(id = id)
        pets[id] = updated
        return updated
    }

    override fun findPetsByStatus(status: String): List<Pet> =
        pets.values.filter { it.status?.value == status }

    override fun findPetsByTags(tags: List<String>): List<Pet> =
        pets.values.filter { pet -> pet.tags?.any { tag -> tags.contains(tag.name) } == true }

    override fun getPetById(petId: Long): Pet? = pets[petId]

    override fun updatePetWithForm(petId: Long, name: String?, status: String?): Pet {
        val existing = pets[petId] ?: throw NotFoundException("Pet not found: $petId")
        val updated = existing.copy(
            name = name ?: existing.name,
            status = status?.let { s -> Pet.Status.entries.find { it.value == s } } ?: existing.status
        )
        pets[petId] = updated
        return updated
    }

    override fun deletePet(petId: Long) {
        pets.remove(petId) ?: throw NotFoundException("Pet not found: $petId")
    }

    override fun uploadFile(petId: Long, additionalMetadata: String?, bytes: ByteArray): ModelApiResponse {
        pets[petId] ?: throw NotFoundException("Pet not found: $petId")
        val id = nextPhotoId()
        photos[id] = bytes
        return ModelApiResponse(code = 200, type = "unknown", message = "File uploaded: ${bytes.size} bytes")
    }

    // -------------------------------------------------------------------------
    // Store
    // -------------------------------------------------------------------------

    override fun getInventory(): Map<String, Int> {
        val result = mutableMapOf<String, Int>()
        for (pet in pets.values) {
            val s = pet.status?.value ?: continue
            result[s] = (result[s] ?: 0) + 1
        }
        return result
    }

    override fun placeOrder(order: Order): Order {
        val id = order.id ?: nextOrderId()
        val saved = order.copy(id = id)
        orders[id] = saved
        return saved
    }

    override fun getOrderById(orderId: Long): Order? = orders[orderId]

    override fun deleteOrder(orderId: Long) {
        orders.remove(orderId) ?: throw NotFoundException("Order not found: $orderId")
    }

    // -------------------------------------------------------------------------
    // User
    // -------------------------------------------------------------------------

    override fun createUser(user: User): User {
        val uname = user.username ?: throw InvalidInputException("Username required")
        val id = user.id ?: ((users.values.mapNotNull { it.id }.maxOrNull() ?: 0L) + 1L)
        val saved = user.copy(id = id)
        users[uname] = saved
        return saved
    }

    override fun createUsersWithList(users: List<User>): User {
        for (user in users) createUser(user)
        return getUserByName(users.first().username ?: "")!!
    }

    override fun loginUser(username: String, password: String): String {
        users[username] ?: throw NotFoundException("User not found: $username")
        return "logged-in-$username"
    }

    override fun logoutUser() { /* stateless no-op */ }

    override fun getUserByName(username: String): User? = users[username]

    override fun updateUser(username: String, user: User) {
        users[username] ?: throw NotFoundException("User not found: $username")
        users[username] = user.copy(username = username)
    }

    override fun deleteUser(username: String) {
        users.remove(username) ?: throw NotFoundException("User not found: $username")
    }
}

File: ruby/rails/lib/in_memory_petstore_repository.rb
require 'json'
require_relative 'petstore_errors'

# In-memory Petstore repository for use in tests.
# Mirrors the behaviour of PostgresPetstoreRepository without any DB connection.
class InMemoryPetstoreRepository
  def initialize
    @pets   = {}
    @orders = {}
    @users  = {}
    @photos = {}
  end

  # ---------------------------------------------------------------------------
  # Pet operations
  # ---------------------------------------------------------------------------

  def add_pet(pet)
    id = fetch(pet, :id, 'id')&.to_i || next_id(@pets)
    data = normalize_pet(pet.merge('id' => id))
    @pets[id] = data
    data
  end

  def update_pet(pet)
    id = fetch(pet, :id, 'id')&.to_i
    raise InvalidInputError, 'Pet ID required for update' if id.nil?
    raise NotFoundError, 'Pet not found' unless @pets.key?(id)

    data = normalize_pet(pet)
    @pets[id] = data
    data
  end

  def find_pet_by_id(pet_id)
    @pets.fetch(pet_id.to_i) { raise NotFoundError, 'Pet not found' }
  end

  def delete_pet(pet_id)
    @pets.delete(pet_id.to_i)
  end

  def find_pets_by_status(status)
    @pets.values.select { |p| p['status'] == status }
  end

  def find_pets_by_tags(tags)
    return [] if tags.nil? || tags.empty?

    @pets.values.select do |p|
      pet_tag_names = (p['tags'] || []).map { |t| t['name'] }
      tags.any? { |t| pet_tag_names.include?(t) }
    end
  end

  def update_pet_with_form(pet_id, name, status)
    pet = @pets.fetch(pet_id.to_i) { raise NotFoundError, 'Pet not found' }
    pet['name']   = name   if name
    pet['status'] = status if status
  end

  def upload_file(pet_id, content, content_type, metadata)
    raise NotFoundError, 'Pet not found' unless @pets.key?(pet_id.to_i)

    id = next_id(@photos)
    @photos[id] = {
      'id'          => id,
      'petId'       => pet_id.to_i,
      'content'     => content,
      'contentType' => content_type,
      'metadata'    => metadata
    }
    content.bytesize
  end

  # ---------------------------------------------------------------------------
  # Store operations
  # ---------------------------------------------------------------------------

  def get_inventory
    @pets.values.each_with_object({}) do |pet, inv|
      s = pet['status']
      next unless s
      inv[s] = (inv[s] || 0) + 1
    end
  end

  def place_order(order)
    id = fetch(order, :id, 'id')&.to_i || next_id(@orders)
    data = normalize_order(order.merge('id' => id))
    @orders[id] = data
    data
  end

  def find_order_by_id(order_id)
    @orders.fetch(order_id.to_i) { raise NotFoundError, 'Order not found' }
  end

  def delete_order(order_id)
    @orders.delete(order_id.to_i)
  end

  # ---------------------------------------------------------------------------
  # User operations
  # ---------------------------------------------------------------------------

  def create_user(user)
    username = fetch(user, :username, 'username')
    @users[username] = normalize_user(user)
    user
  end

  def create_users_with_list(users)
    users.each { |u| create_user(u) }
  end

  def find_user_by_username(username)
    @users.fetch(username) { raise NotFoundError, 'User not found' }
  end

  def update_user(username, user)
    raise NotFoundError, 'User not found' unless @users.key?(username)

    @users[username] = normalize_user(user.merge('username' => username))
  end

  def delete_user(username)
    @users.delete(username)
  end

  def login_user(username, password)
    user = @users[username]
    return false unless user

    user['password'] == password
  end

  private

  def fetch(obj, sym_key, str_key) = obj[sym_key] || obj[str_key]
  def next_id(store)                = (store.keys.max || 0) + 1

  def normalize_pet(pet)
    id = fetch(pet, :id, 'id')&.to_i
    {
      'id'        => id,
      'name'      => fetch(pet, :name, 'name'),
      'category'  => fetch(pet, :category, 'category'),
      'photoUrls' => fetch(pet, :photoUrls, 'photoUrls') || fetch(pet, :photo_urls, 'photo_urls') || [],
      'tags'      => fetch(pet, :tags, 'tags'),
      'status'    => fetch(pet, :status, 'status')
    }.compact.tap { |h| h['photoUrls'] ||= [] }
  end

  def normalize_order(order)
    id = fetch(order, :id, 'id')&.to_i
    pet_id_val = fetch(order, :petId, 'petId') || fetch(order, :pet_id, 'pet_id')
    qty = fetch(order, :quantity, 'quantity')
    {
      'id'       => id,
      'petId'    => pet_id_val&.to_i,
      'quantity' => qty&.to_i,
      'shipDate' => fetch(order, :shipDate, 'shipDate') || fetch(order, :ship_date, 'ship_date'),
      'status'   => fetch(order, :status, 'status'),
      'complete' => fetch(order, :complete, 'complete')
    }.compact
  end

  def normalize_user(user)
    id_val = fetch(user, :id, 'id')
    status_val = fetch(user, :userStatus, 'userStatus') || fetch(user, :user_status, 'user_status')
    {
      'id'         => id_val&.to_i,
      'username'   => fetch(user, :username, 'username'),
      'firstName'  => fetch(user, :firstName, 'firstName') || fetch(user, :first_name, 'first_name'),
      'lastName'   => fetch(user, :lastName, 'lastName') || fetch(user, :last_name, 'last_name'),
      'email'      => fetch(user, :email, 'email'),
      'password'   => fetch(user, :password, 'password'),
      'phone'      => fetch(user, :phone, 'phone'),
      'userStatus' => status_val&.to_i
    }.compact
  end
end
